### ISSUE TITLE
dsPIC30 ELF relocation issue

### ISSUE THREAD

**Issue by ryanmkurtz** [NationalSecurityAgency/ghidra#1689]
Here is a small program's ELF object file for the dsPIC30:

```
#include <stdio.h>

int main(void) {

    printf("Hello World\n");
}
```
<img width="500" alt="Capture" src="https://user-images.githubusercontent.com/5378554/77940970-71959000-7287-11ea-8247-6a9c35135883.PNG">

Here is objdump's relocation information for the `.text` section:
```
OFFSET   TYPE              VALUE
00000002 WORD              .LC0
00000004 PC RELATIVE BRANCH  _puts
```
And Ghidra's corresponding relocation table entry:
<img width="438" alt="Captu2re" src="https://user-images.githubusercontent.com/5378554/77941360-0a2c1000-7288-11ea-9c2c-824f380746d1.PNG">

Ghidra appears to be fixing up relocations in the wrong locations, resulting in a bad call target and a corrupted return instruction. 

[main.zip](https://github.com/NationalSecurityAgency/ghidra/files/4404371/main.zip)

**Comment by ghidra1** [comment]
Could you please provide a fully linked binary which includes this object module.  Need to ensure that while addressing object module concerns we do not break those for a fully linked module.

**Comment by ryanmkurtz** [comment]
Sure:

[Test.X.production.zip](https://github.com/NationalSecurityAgency/ghidra/files/4416741/Test.X.production.zip)

**Comment by ghidra1** [comment]
There are quite a few issues:
1. relocation offset needs to be treated as byte offset within relevant section

**Comment by ryanmkurtz** [comment]
Additionally, the object file for this code introduces two new relocation types that aren't implemented: **0x15** and **0x18**.

```
__prog__ const char __attribute__((space(psv))) hello_world[] = "Hello World";

int main(void) {

    printf("%s\n", hello_world);
    printf("Hello World!\n");
}
```

[main.zip](https://github.com/NationalSecurityAgency/ghidra/files/4427143/main.zip)

**Comment by ryanmkurtz** [comment]
```
__prog__ const char __attribute__((space(psv))) hello_world[] = "Hello World";

int main(void) {

    printf("%s, %d\n", hello_world, 5);
    printf("Hello World, %d\n", 6);
}
```
[main.zip](https://github.com/NationalSecurityAgency/ghidra/files/4428392/main.zip)
[Test.X.zip](https://github.com/NationalSecurityAgency/ghidra/files/4428393/Test.X.zip)

**Comment by ryanmkurtz** [comment]
```
__prog__ const char __attribute__((space(psv))) hello_world[] = "Hello World";


int sum(int n, ...) {
    int i, ret = 0;
    va_list valist;
    va_start(valist, n);
    
    for (i = 0; i < n; i++) {
        ret += va_arg(valist, int);
    }
    
    va_end(valist);
    
    return ret;
}

int main(void) {

    printf("%s, %d\n", hello_world, 5);
    printf("Hello World, %d\n", 6);
    printf("sum(5, 1,2,3,4,5) = %d", sum(5,1,2,3,4,5));
}
```
[main.zip](https://github.com/NationalSecurityAgency/ghidra/files/4438123/main.zip)
[Test.X.zip](https://github.com/NationalSecurityAgency/ghidra/files/4438125/Test.X.zip)

**Comment by ryanmkurtz** [comment]
Fixed by 305a1ddf98076cfe540e1d7046dc774fd6b89a67

**Comment by emteere** [comment]
Fixed in master and future 9.2
